### ISSUE TITLE
Non-negativity constraints

### ISSUE THREAD

**Issue by david-cortes** [rexyai/rsparse#36]
In `WRMF`, there is an option `non_negative`, which is handled by first finding the unconstrained factors and then setting them to zero if they turn out negative:
```r
# if need non-negative matrix factorization - just set all negative values to zero
if (private$non_negative) private$U[private$U < 0] = 0
```

But in general, the solution of a constrained linear system is not the solution of the unconstrained system with the non-conforming coefficients set to the bounds (see e.g. https://en.wikipedia.org/wiki/Non-negative_least_squares).

Now, since it iterates multiple times, the fitted factors are probably not going to be too bad (save for the last iteration), but then it again does the same thing when calling `transform`, which is not optimal.

Did this technique come from some reference?

**Comment by dselivanov** [comment]
Good catch about transform!

As for method - yes it is a NNLS workaround.

I didn't find goog nnls solver at the time of writing (do you know any
C/C++ solver which is easy to embed?). So I decided to try approach similar
to "projected gradient" which essentially means setting negative values to
zero at each step.

**Comment by david-cortes** [comment]
Not sure about NNLS optimizers, and am not familiar with the state-of-the-art in non-negative matrix factorization.

Perhaps it’s an overkill, but it’s possible to use a generalized optimizer for differentiable functions with bound constraints – for example, base R provides a C port of the L-BFGS-B solver (the header is in `Rext/Applic.h`), but it uses R’s allocators so it cannot be used in an openmp parallel section. There’s also other C/C++ ports of L-BFGS-B, such as [this](https://github.com/yixuan/LBFGSpp), and other good optimizers such as  [TNCG](https://github.com/scipy/scipy/blob/master/scipy/optimize/tnc/tnc.c) (modified version [here](https://github.com/david-cortes/poismf/blob/master/src/tnc.c) for using it in parallel), or [conjugate gradient methods](https://github.com/david-cortes/poismf/blob/master/src/nonnegcg.c), but from my experiments, using a bounded gradient-based procedure can end up being an order of magnitude slower than the unconstrained version using Cholesky.

Perhaps a better option could be to take the NNLS solver from SciPy, which is a stand-alone Fortran file: https://github.com/scipy/scipy/blob/master/scipy/optimize/__nnls/nnls.f
It looks very easy to use, although I don’t think it’s the fastest solver out there, especially since it’s not specialized for symmetric matrices.

Taking a look at other software, it looks like they use more elaborate procedures though – for example, scikit-learn lists these as references:
https://arxiv.org/pdf/1010.1763.pdf
http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.214.6398&rep=rep1&type=pdf
(second one is easier and looks more similar to what you were doing)

And the one from Julia seems to be based on proximal gradients (looks very inefficient though).

**Comment by dselivanov** [comment]
@david-cortes I've added NNLS solver from https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.1075.4794&rep=rep1&type=pdf. Code adapted from https://github.com/linxihui/NNLM/. 
Let's see how it will work as subroutine in WRMF.

**Comment by david-cortes** [comment]
I think it might be better to split `WRMF` between the implicit feedback variant and the explicit feedback variant (which is not weighted) so as to make the available options for each more clear.

Also perhaps could be a good opportunity to add an option in the explicit variant for biases and mean centering too - it shouldn't require too many modifications.

**Comment by david-cortes** [comment]
Actually, looking at the paper again, I see the same method could also be applied to the implicit feedback variant with minor modifications. Seems spark also offers a non-negative implicit feedback variant: https://spark.apache.org/docs/latest/api/python/_modules/pyspark/mllib/recommendation.html#ALS.trainImplicit

**Comment by dselivanov** [comment]
> Actually, looking at the paper again, I see the same method could also be applied to the implicit feedback variant

Yes that is the plan. We just needed a non-negative variant of `solve`  which can be used from C++ code inside threads. `nnls` solver from my commit is not that fast as `nnls::nnls` (which is significantly slower than cholesky). Hopefully will share some results in coming days.

**Comment by dselivanov** [comment]
@david-cortes take a look into non-negative solver for implicit feedback. It seems it is only a little bit slower than Cholesky. In terms of metrics it seems it noticeably worse compared to one without non-negative constraints.

**Comment by david-cortes** [comment]
Seems to be working correctly, and the metrics are better than when the earlier strategy, although they are indeed bad compared to the unconstrained version.

**Comment by david-cortes** [comment]
Actually, upon further inspection, it's not giving the exact solution, will try to post an example but it will involve some python.

**Comment by david-cortes** [comment]
Well had some errors with my test, in the end it **is** working correctly, up to some decimal precision. Here I'm leaving a short test:
```r
library(Matrix)
library(rsparse)
library(reticulate)

data("movielens100k")
X = movielens100k

set.seed(123)
mf = WRMF$new(rank=50, lambda=1, non_negative=TRUE)
temp = mf$fit_transform(X)
rm(temp)
ItemFactors = t(mf$components)
chosen_user = 1L
user_vec = as.numeric(mf$transform(X[chosen_user, , drop=FALSE]))

x_user = as(X[chosen_user, , drop=FALSE], "RsparseMatrix")
user_ix = x_user@j
user_val = x_user@x
x_dense = rep(1, ncol(X))
x_dense[user_ix + 1L] = user_val
x_bin = as.numeric(as.numeric(x_user) > 0)


loss_func <- function(user_vec, lambda=1) {
  pred = ItemFactors %*% user_vec
  err = x_dense * (pred - x_bin)^2
  res = sum(err) + lambda*(user_vec %*% user_vec)
  return(as.numeric(res))
}

loss_func(user_vec)
# [1] 245.1464

fun_for_py = "
import numpy as np
from scipy.optimize import minimize
def fun(x, *args):
    ItemFactors_py, x_bin_py, x_dense_py, lambda_py = args
    pred = ItemFactors_py.dot(x)
    err = (pred - x_bin_py) ** 2
    err = np.sum(x_dense_py * err)
    return err + lambda_py*x.dot(x)
def get_sol(ItemFactors_py, x_bin_py, x_dense_py, lambda_py):
    k = ItemFactors_py.shape[1]
    args_pass = (ItemFactors_py, x_bin_py, x_dense_py, lambda_py)
    return minimize(
        fun, np.zeros(k), args_pass, bounds=[(0.,np.inf) for i in range(k)], method='TNC'
    )['x']
"
temp_file = file.path(tempdir(), "pyfun.py")
fileConn = file(temp_file)
writeLines(fun_for_py, fileConn)
close(fileConn)
source_python(temp_file)

np = import("numpy")
ItemFactors_py = np$array(ItemFactors)
x_bin_py = np$array(x_bin)
x_dense_py = np$array(x_dense)
lambda_py = 1.
loss_func(get_sol(ItemFactors_py, x_bin_py, x_dense_py, lambda_py))
# [1] 244.8368
```
